File: pbworkspace/pbworkspace-jar/src/main/java/it/csi/pbandi/pbworkspace/business/service/NotificheViaMailService.java
/*******************************************************************************
* Copyright Regione Piemonte - 2024
* SPDX-License-Identifier: EUPL-1.2
******************************************************************************/

package it.csi.pbandi.pbworkspace.business.service;

import java.security.InvalidParameterException;
import java.util.List;

import javax.servlet.http.HttpServletRequest;
import javax.ws.rs.Consumes;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import org.jboss.resteasy.annotations.jaxrs.QueryParam;

import it.csi.pbandi.pbworkspace.exception.ErroreGestitoException;
import it.csi.pbandi.pbworkspace.exception.UtenteException;
import it.csi.pbandi.pbworkspace.integration.vo.NotificaAlertVO;

@Path("/notificheViaMail")
public interface NotificheViaMailService {

	@GET
	@Path("getMail")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getMail(@Context HttpServletRequest req) throws UtenteException, Exception;

	@POST
	@Path("saveMail")
	@Consumes(MediaType.APPLICATION_JSON)
	@Produces(MediaType.APPLICATION_JSON)
	public Response saveMail(@Context HttpServletRequest req, String mail)
			throws UtenteException, InvalidParameterException, Exception;

	@GET
	@Path("getNotificheFrequenze")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getNotificheFrequenze(@Context HttpServletRequest req) throws UtenteException, Exception;

	@POST
	@Path("associateNotificheIstruttore")
	@Consumes(MediaType.APPLICATION_JSON)
	@Produces(MediaType.APPLICATION_JSON)
	public Response associateNotificheIstruttore(@Context HttpServletRequest req, List<NotificaAlertVO> notificheAlert)
			throws UtenteException, InvalidParameterException, Exception;

	@GET
	@Path("getBandiProgetti")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getBandiProgetti(@Context HttpServletRequest req,
			@QueryParam("idSoggettoNotifica") Long idSoggettoNotifica) throws ErroreGestitoException, UtenteException;

	@POST
	@Path("associateProgettiToNotifica")
	@Consumes(MediaType.APPLICATION_JSON)
	@Produces(MediaType.APPLICATION_JSON)
	public Response associateProgettiToNotifica(@Context HttpServletRequest req, NotificaAlertVO notificaAlert)
			throws UtenteException, InvalidParameterException, Exception;

}

File: pbworkspacewcl/src/app/notifiche-via-mail/services/notifiche-via-mail.service.ts
/*
 * Copyright Regione Piemonte - 2024 
 * SPDX-License-Identifier: EUPL-1.2 
*/

import { HttpClient, HttpHeaders, HttpParams } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { ConfigService } from "src/app/core/services/config.service";
import { HandleExceptionService } from "src/app/core/services/handle-exception.service";
import { BandoProgettiVO } from "../commons/vo/bando-progetti-vo";
import { NotificaAlertVO } from "../commons/vo/notifica-alert-vo";
import { NotificheFrequenzeVO } from "../commons/vo/notifiche-frequenze-vo";

@Injectable()
export class NotificheViaMailService {

    constructor(
        private configService: ConfigService,
        private http: HttpClient,
        private handleExceptionService: HandleExceptionService
    ) { }

    getMail(): Observable<string> {
        let url = this.configService.getApiURL() + "/restfacade/notificheViaMail/getMail";
        return this.http.get(url, { responseType: 'text' });
    }

    saveMail(email: string) {
        let url = this.configService.getApiURL() + "/restfacade/notificheViaMail/saveMail";
        let headers = new HttpHeaders().set('Content-Type', 'application/json');
        return this.http.post<boolean>(url, email, { headers: headers });
    }

    getNotificheFrequenze(): Observable<NotificheFrequenzeVO> {
        let url = this.configService.getApiURL() + "/restfacade/notificheViaMail/getNotificheFrequenze";
        return this.http.get<NotificheFrequenzeVO>(url);
    }

    associateNotificheIstruttore(notificheAlert: Array<NotificaAlertVO>) {
        let url = this.configService.getApiURL() + "/restfacade/notificheViaMail/associateNotificheIstruttore";
        let headers = new HttpHeaders().set('Content-Type', 'application/json');
        return this.http.post<boolean>(url, notificheAlert, { headers: headers });
    }

    getBandiProgetti(idSoggettoNotifica: number): Observable<Array<BandoProgettiVO>> {
        let url = this.configService.getApiURL() + "/restfacade/notificheViaMail/getBandiProgetti";
        let params = new HttpParams().set("idSoggettoNotifica", idSoggettoNotifica.toString());
        return this.http.get<Array<BandoProgettiVO>>(url, { params: params });
    }

    associateProgettiToNotifica(notificaAlert: NotificaAlertVO) {
        let url = this.configService.getApiURL() + "/restfacade/notificheViaMail/associateProgettiToNotifica";
        let headers = new HttpHeaders().set('Content-Type', 'application/json');
        return this.http.post<boolean>(url, notificaAlert, { headers: headers });
    }

}
